lua/ttt2/libraries/door.lua

```diff
@@ -0,0 +1,606 @@
+			return not ent:HasSpawnFlags(SF_PROP_DOOR_USE_CLOSES)
+		elseif door.IsValidSpecial(cls) then
+			return not ent:HasSpawnFlags(SF_FUNC_DOOR_MODE_TOGGLE)
+		end
+
+		return false
+	end
+
+	---
+	-- Returns if a door is destructible by reading out the internal variable.
+	-- @param Entity ent The door entity that should be checked
+	-- @return boolean Returns if the door is destructible
+	-- @internal
+	-- @realm server
+	function door.IsDestructible(ent)
+		local cls = ent:GetClass()
+
+		if door.IsValidNormal(cls) then
+			return ent:HasSpawnFlags(SF_PROP_DOOR_START_BREAKABLE)
+		elseif door.IsValidSpecial(cls) then
+			return false
+		end
+
+		return false
+	end
+
+	---
+	-- Sets if a player can use interact with a door by setting the internal variable.
+	-- @param Entity ent The door entity that should be checked
+	-- @param boolean state The new player use state
+	-- @internal
+	-- @realm server
+	function door.SetPlayerCanUse(ent, state)
+		local cls = ent:GetClass()
+
+		if door.IsValidNormal(cls) then
+			if state then
+				RemoveSpawnFlag(ent, SF_PROP_DOOR_IGNORE_PLAYER_USE)
+			else
+				AddSpawnFlag(ent, SF_PROP_DOOR_IGNORE_PLAYER_USE)
+			end
+		elseif door.IsValidSpecial(cls) then
+			-- 256: use opens
+			if state then
+				AddSpawnFlag(ent, SF_FUNC_DOOR_PLAYER_USE_OPENS)
+			else
+				RemoveSpawnFlag(ent, SF_FUNC_DOOR_PLAYER_USE_OPENS)
+			end
+		end
+	end
+
+	---
+	-- Sets if a player can touch interact with a door by setting the internal variable.
+	-- @param Entity ent The door entity that should be checked
+	-- @param boolean state The new player touch state
+	-- @internal
+	-- @realm server
+	function door.SetPlayerCanTouch(ent, state)
+		local cls = ent:GetClass()
+
+		if door.IsValidSpecial(cls) then
+			-- 1024: touch opens
+			if state then
+				AddSpawnFlag(ent, SF_FUNC_DOOR_TOUCH_OPENS)
+			else
+				RemoveSpawnFlag(ent, SF_FUNC_DOOR_TOUCH_OPENS)
+			end
+		end
+	end
+
+	---
+	-- Sets if a door autocloses by setting the internal variable.
+	-- @param Entity ent The door entity that should be checked
+	-- @param boolean state The new autoclose state
+	-- @internal
+	-- @realm server
+	function door.SetAutoClose(ent, state)
+		local cls = ent:GetClass()
+
+		if door.IsValidNormal(cls) then
+			-- 8192: door closes on use
+			if state then
+				RemoveSpawnFlag(ent, SF_PROP_DOOR_USE_CLOSES)
+				ent:SetKeyValue("returndelay", 3)
+			else
+				AddSpawnFlag(ent, SF_PROP_DOOR_USE_CLOSES)
+			end
+		elseif door.IsValidSpecial(cls) then
+			if state then
+				RemoveSpawnFlag(ent, SF_FUNC_DOOR_MODE_TOGGLE)
+			else
+				AddSpawnFlag(ent, SF_FUNC_DOOR_MODE_TOGGLE)
+			end
+		end
+	end
+
+	---
+	-- Handles the damage of doors that are still in the wall.
+	-- Called in @{GM:EntityTakeDamage}.
+	-- @param Entity ent The entity that is damages
+	-- @param CTakeDamageInfo dmginfo The damage info object
+	-- @internal
+	-- @realm server
+	function door.HandleDamage(ent, dmginfo)
+		if not ent:DoorIsDestructible() then return end
```

Comment:
Since sh_door is an entity extension all entites have this function as far as I know. Additionally it is an internal function that shouldn't be called on something else.